lib/srv/heartbeatv2.go

```diff
@@ -73,20 +74,32 @@ func NewSSHServerHeartbeat(cfg SSHServerHeartbeatConfig) (*HeartbeatV2, error) {
 		return nil, trace.Wrap(err)
 	}
 
+	var metadataPtr atomic.Pointer[metadata.Metadata]
 	inner := &sshServerHeartbeatV2{
-		getServer: func(ctx context.Context) *types.ServerV2 {
-			server := cfg.GetServer()
-			metadata, err := metadata.Get(ctx)
-			if err == nil {
-				if metadata.CloudMetadata != nil {
-					server.SetCloudMetadata(metadata.CloudMetadata)
+		getMetadata: metadata.Get,
+		announcer:   cfg.Announcer,
+	}
+	inner.getServer = func(ctx context.Context) *types.ServerV2 {
+		server := cfg.GetServer()
+
+		meta := metadataPtr.Load()
+		if meta == nil {
+			go func() {
+				meta, err := inner.getMetadata(ctx)
+				if err != nil {
+					log.Warnf("Failed to get metadata: %v", err)
+				} else if meta != nil && meta.CloudMetadata != nil {
+					// Set the metadata immediately to give the heartbeat
+					// a chance to use it.
+					server.SetCloudMetadata(meta.CloudMetadata)
+					metadataPtr.CompareAndSwap(nil, meta)
```

Comment:
This will take the first result, not the most recent result. Probably fine, just want to confirm this is what you intend. Consider the following case:

- heartbeat 1 observes nil metadata, kicks off background goroutine
- background routine still in progress, heartbeat 2 also sees nil metadata and kicks off its own background routine
- background routine 1 completes, the compare and swap sets meta
- background routine 2 completes, result is ignored

Reply:
That's fine. `metadata.Get` is cached, so the result would be the same no matter which result we use.